### ISSUE TITLE
False-positive `reportOverlappingOverload`

### ISSUE THREAD

**Issue by saryou** [microsoft/pyright#10499]
Starting from 1.1.394, `reportOverlappingOverload` is being incorrectly triggered.

## Code to Reproduce

```python
from typing import overload
from datetime import datetime, date

# ✅ Works as expected
@overload
def test_1(a: int) -> int:
    ...

@overload
def test_1(a: float) -> float:
    ...

def test_1(a: int | float) -> int | float:
    return a

# information: Type of "test_1(1)" is "int"
reveal_type(test_1(1))
# information: Type of "test_1(1.1)" is "float"
reveal_type(test_1(1.1))


# ❌ error: Overload 1 for "test_2" overlaps overload 2 and returns an incompatible type (reportOverlappingOverload)
@overload
def test_2(a: int) -> list[int]:
    ...

@overload
def test_2(a: float) -> list[float]:
    ...

def test_2(a: int | float) -> list[int] | list[float]:
    return [a]

# information: Type of "test_2(1)" is "list[int]"
reveal_type(test_2(1))
# information: Type of "test_2(1.1)" is "list[float]"
reveal_type(test_2(1.1))


# ❌ error: Overload 1 for "test_3" overlaps overload 2 and returns an incompatible type (reportOverlappingOverload)
@overload
def test_3(a: datetime) -> list[datetime]:
    ...

@overload
def test_3(a: date) -> list[date]:
    ...

def test_3(a: date | datetime) -> list[date] | list[datetime]:
    return [a]

# information: Type of "test_3(datetime.min)" is "list[datetime]"
reveal_type(test_3(datetime.min))
# information: Type of "test_3(date.min)" is "list[date]"
reveal_type(test_3(date.min))

# NOTE: If the order of the overloads for test_3 is reversed, pyright says:
#   error: Overload 2 for "test_3" will never be used because its parameters overlap overload 1 (reportOverlappingOverload)
#   information: Type of "test_3(datetime.min)" is "list[date]"
#   information: Type of "test_3(date.min)" is "list[date]"
```

**Comment by erictraut** [comment]
This isn't a false positive. Pyright is working correctly here, so this isn't a bug.

An "overlapping overload" error is reported when you can pass the same set of arguments to two or more overloads and the return type of the later overload is not a subtype of the return type of the earlier overload. That's precisely the condition you see with your code. This can lead to unsoundness. Let's use your second example to demonstrate the problem that can occur with this type of a bug.

We can pass a value of `1` (an instance of `int`) to `test_2`, but depending on how we pass it, the call evaluates to either `list[int]` or `list[float]`. Those two types have no overlap, so it's not possible for the _actual_ return type to be compatible with both of these types. 

```python
def as_float(x: int) -> float:
    return x

reveal_type(test_2(1)) # list[int]
reveal_type(test_2(as_float(1)) # list[float]
```

If you don't want pyright to warn you about invalid overlapping overloads, you can disable the `reportOverlappingOverload` check. Or you can suppress the error on a case-by-case basis using a `# pyright: ignore` comment.

**Comment by erictraut** [comment]
In the Python type system, `type[int]` is a subtype of `type[float]`. According to [this section of the spec](https://typing.python.org/en/latest/spec/special-types.html#type), the type parameter of `type` is covariant. Since `int` is a subtype of `float` (according to [this section of the spec](https://typing.python.org/en/latest/spec/special-types.html#special-cases-for-float-and-complex)), that means `type[int]` is a subtype of `type[float]`. This makes sense when you consider that the annotation `float` is interpreted as a union of `float` and `int`, so `type[float]` is equivalent to `type[float | int]` which is equivalent to `type[float] | type[int]` (since the spec says `type` distributes over unions). Obviously, `type[int]` is a subtype of `type[float] | type[int]`, which means it's also a subtype of `type[float]`.

**Comment by erictraut** [comment]
Yeah, any time `float` appears in a type expression (e.g. an annotation), a type checker effectively treats it as `float | int`. There's currently no way in the Python type system to say "I just want a `float`, not an `int`". There has been [some discussion](https://discuss.python.org/t/options-for-a-long-term-fix-of-the-special-case-for-float-int-complex/54289) recently about remedying this (e.g. by adding a special form such as `typing.Float` to represent this concept), but the community hasn't reached consensus yet.
